### PR TITLE
Generalize with variance

### LINKED ISSUE #667: Endless loop in recursive solver
```rust
#[test]
fn endless_loop() {
    test! {
        disable_coherence;
        program {
            trait FnOnce {
                type Output;
            }

            struct MyClosure<F> {}
            impl<T> FnOnce for MyClosure<fn() -> T> {
                type Output = T;
            }
        }
        goal {
            exists<T> {
                <MyClosure<fn() -> T> as FnOnce>::Output = T
            }
        } yields[SolverChoice::recursive_default()] {
            "Unique; for<?U0> { substitution [?0 := ^0.0], lifetime constraints [] }"
        }
    }
}
```
This hangs forever in the recursive solver; in the SLG solver, it results in `Ambiguous; no inference guidance`, which seems wrong as well? Replacing `fn() -> T` by `T` makes the test succeed (in both solvers).

CC rust-analyzer/rust-analyzer#6945.

### PR DESCRIPTION
**Pull request by jackh726**
Closes #667

### CODE REVIEW

File: chalk-solve/src/infer/unify.rs [hunk 1]
```diff
@@ -660,7 +740,9 @@ impl<'t, I: Interner> Unifier<'t, I> {
                 if matches!(kind, TyVariableKind::Integer | TyVariableKind::Float) {
                     ty.clone()
                 } else if let Some(ty) = self.table.normalize_ty_shallow(interner, ty) {
-                    self.generalize_ty(&ty, universe_index)
+                    self.generalize_ty(&ty, universe_index, variance)
+                } else if matches!(variance, Variance::Invariant) {
```
Comment (matthewjasper) [inline chalk-solve/src/infer/unify.rs:744]: The rustc code here also checks that this variable can name `universe_index` here, although that's maybe only for error messages.

File: chalk-solve/src/infer/unify.rs [hunk 2]
```diff
@@ -674,6 +756,7 @@ impl<'t, I: Interner> Unifier<'t, I> {
         &mut self,
         lifetime: &Lifetime<I>,
         universe_index: UniverseIndex,
+        variance: Variance,
     ) -> Lifetime<I> {
```
Comment (matthewjasper) [inline chalk-solve/src/infer/unify.rs:760]: Lifetimes are also cloned in the `Invariant` case.

File: chalk-solve/src/infer/unify.rs (hunk 1 above)
Comment (jackh726) [reply chalk-solve/src/infer/unify.rs:744]: Hmm. So should I change this here?

File: chalk-solve/src/infer/unify.rs (hunk 2 above)
Comment (jackh726) [reply chalk-solve/src/infer/unify.rs:760]: Updated.

File: chalk-solve/src/infer/unify.rs (hunk 1 above)
Comment (matthewjasper) [reply chalk-solve/src/infer/unify.rs:744]: I think that we can keep the simpler code for now.

**Comment by lnicola** [comment]
If it's not too much of a hassle, would it be possible to publish a new version to crates.io?

**Comment by jackh726** [comment]
@lnicola just kicked off a publish

**Comment by lnicola** [comment]
Note to self: we should do `chalk` upgrades on Monday.

**Pull request merged**